core/cli/node/client.hpp

```diff
@@ -0,0 +1,671 @@
+                   "provider to use for retrieval, if not present it'll use "
+                   "local discovery",
+                   Address)
+      provider;
+      CLI_OPTIONAL("pieceCid",
+                   "require data to be retrieved from a specific Piece CID",
+                   CID)
+      piece_cid;
+      CLI_DEFAULT(
+          "maxPrice",
+          fmt::format(
+              "maximum price the client is willing to consider (default: {}"
+              " FIL)",
+              kDefaultMaxRetrievePrice.atto())
+              .c_str(),
+          AttoFil,
+          {kDefaultMaxRetrievePrice})
+      max_price;
+      CLI_OPTIONAL("data-selector",
+                   "(NOT SUPPORTED) IPLD datamodel text-path selector, or IPLD "
+                   "json selector",
+                   std::string)
+      data_selector;
+      CLI_BOOL("car", "Export to a car file instead of a regular file") car;
+      CLI_BOOL("allow-local", "allow use local imports") allow_local;
+      CLI_BOOL("car-export-merkle-proof",
+               "(requires --data-selector and --car) Export data-selector "
+               "merkle proof")
+      car_export_merkle_proof;
+
+      CLI_OPTS() {
+        Opts opts;
+        from(opts);
+        provider(opts);
+        piece_cid(opts);
+        allow_local(opts);
+        car(opts);
+        max_price(opts);
+        data_selector(opts);
+        car_export_merkle_proof(opts);
+        return opts;
+      }
+    };
+
+    CLI_RUN() {
+      const Node::Api api{argm};
+
+      const CID data_cid{cliArgv<CID>(argv, 0, "dataCid")};
+      const std::string path{cliArgv<std::string>(argv, 1, "path")};
+
+      const boost::optional<CID> piece_cid = args.piece_cid.v;
+      RetrievalOrder order{
+          .client =
+              (args.from ? *args.from
+                         : cliTry(api->WalletDefaultAddress(),
+                                  "Getting address of default wallet...")),
+          .miner = (args.provider ? *args.provider : Address{})};
+
+      fmt::print("max price is {} fil ({} attofil)\n",
+                 args.max_price->fil,
+                 args.max_price->atto());
+
+      fmt::print("retrieving {} to {} {}\n",
+                 data_cid,
+                 args.car ? "car" : "file",
+                 path);
+
+      const FileRef file_ref{.path = path, .is_car = args.car};
+
+      const bool local_found = [&]() {
+        if (args.allow_local) {
+          std::vector<Import> imports = cliTry(api->ClientListImports());
+          for (const Import &import : imports) {
+            if (import.root == data_cid) {
+              fmt::print("Root: {}\n", fmt::to_string(import.root));
+              fmt::print("Local path: {}\n", import.path);
+              return true;
+            }
+          }
+        }
+        return false;
+      }();
+
+      QueryOffer fin_offer;
+      if (not local_found) {  // no local -> make retrieval
+        if (!args.provider) {
```

Comment:
Reads easier
```suggestion
        if (not args.provider.has_value()) {
```

Reply:
fixed

Reply:
.has_value() doesn't exist, but I can "!" replace to "not"